Add support for IP_MULTICAST_LOOP socket option

Abd002 opened the pull request:
net: sockets: Add support for IP_MULTICAST_LOOP socket option

This PR adds support for the IP_MULTICAST_LOOP socket option,
which controls whether multicast packets are looped back to the
sending host.

Fixes: #86977
Signed-off-by: RocketBoy

Abd002 commented:
i can't understand why it failes in twister-build 3, 6

carlescufi commented:
Thanks for the PR! Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

carlescufi commented:
@Abd002 you need to rebase your commits to squash them into a single one

pdgendt commented on include/zephyr/net/net_context.h:
```diff
@@ -205,6 +205,8 @@ struct net_conn_handle;
  * anyway. This saves 12 bytes / context in IPv6.
  */
 __net_socket struct net_context {
+	/* Controls multicast loopback */
+	bool multicast_loop;
```
This should not be the first member, please read the comment below.

jukkar replied:
Place this option to options section like the other options. Also please rename it to `ipv4_multicast_loop`
Also make sure this is only available if `CONFIG_NET_IPV4` is enabled. For example like this.
```
IF_ENABLED(CONFIG_NET_IPV4, (bool ipv4_multicast_loop));
```

sayooj-aerlync replied:
Add the option below the following code:
` 410         /** IPv6 hop limit or IPv4 ttl for packets sent via this context. */
 411         union {
 412                 struct {
 413                         uint8_t ipv6_hop_limit;       /**< IPv6 hop limit */
 414                         uint8_t ipv6_mcast_hop_limit; /**< IPv6 multicast hop limit */
 415                 };
 416                 struct {
 417                         uint8_t ipv4_ttl;       /**< IPv4 TTL */
 418                         uint8_t ipv4_mcast_ttl; /**< IPv4 multicast TTL */
 419                 };
 420         };
 421     
`

sayooj-aerlync replied:
Also let's follow the naming convention used in the file. we can write the variable as `bool ipv4_mcast_loop`

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,56 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
 
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET &&
+	net_ipv4_is_addr_mcast((const struct in_addr *)&NET_IPV4_HDR(pkt)->dst)) {
+			struct net_context *ctx = net_pkt_context(pkt);
+
+		if (ctx && ctx->multicast_loop) {
+			struct net_pkt *clone = net_pkt_clone(pkt, K_NO_WAIT);
+
+			if (clone) {
+				NET_DBG("Looping back IPv4 multicast packet %p", clone);
+	
+				net_pkt_set_iface(clone, net_pkt_iface(pkt));
+	
+				if (net_recv_data(net_pkt_iface(clone), clone) < 0) {
+					NET_ERR("Failed to process cloned IPv4 packet");
+					net_pkt_unref(clone);
+				}
+			} 
+			else {
+					NET_ERR("Failed to clone IPv4 multicast packet");
+			}
+		}
+	}
```
Indentation looks wrong, I'd suggest running clang-format on this block.

pdgendt commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -845,20 +845,41 @@ static inline int z_vrfy_zsock_inet_pton(sa_family_t family,
 #endif
 
 int z_impl_zsock_getsockopt(int sock, int level, int optname,
-			    void *optval, socklen_t *optlen)
+	void* optval, socklen_t* optlen)
```
Don't change formatting here.

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,56 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6 &&
+	net_ipv6_is_addr_mcast((const struct in6_addr *)&NET_IPV6_HDR(pkt)->dst)) {
```
IPv6 is controlled by a separate socket option, `IPV6_MULTICAST_LOOP`. @sayooj-aerlync was looking at that too.

pdgendt reviewed (changes requested):
As Carles suggested, please read the contribution guidelines. I'd expect only 2 commits:

- Introduction of the `IP_MULTICAST_LOOP` definition
- Implementation for `net_context`

jukkar commented on doc/reference/networking/sockets.rst:
```diff
@@ -0,0 +1,8 @@
+Socket Options
+==============
+
+IPv4 Options
+------------
+
+- ``IP_MULTICAST_LOOP``: Controls whether multicast packets are looped back to
+  the sending host. Set to 1 to enable loopback, 0 to disable.
```
This file is not really needed, or if we want to have it it should contain documentation of all the options and not just for this one. Anyway, that is something for another PR.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -2150,6 +2150,7 @@ static struct net_pkt *context_alloc_pkt(struct net_context *context,
 					 size_t len, k_timeout_t timeout)
 {
 	struct net_pkt *pkt;
+	context->multicast_loop = true;
```
Place the default value setting to `net_context_get()` like the other options.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -845,20 +845,41 @@ static inline int z_vrfy_zsock_inet_pton(sa_family_t family,
 
 	SYS_PORT_TRACING_OBJ_FUNC_ENTER(socket, getsockopt, sock, level, optname);
 
+	if (level == IPPROTO_IP && optname == IP_MULTICAST_LOOP) {
+		if (*optlen < sizeof(int)) {
+			errno = EINVAL;
+			ret = -1;
+			goto exit;
+		}
+
+		struct net_context* ctx = zsock_get_context_object(sock);
+		if (ctx == NULL) {
+			errno = EBADF;
+			ret = -1;
+			goto exit;
+		}
+
+		*(int*)optval = ctx->multicast_loop;
+		*optlen = sizeof(int);
+		ret = 0;
+		goto exit;
+	}
+exit:
```
Place the option setting / getting to `subsys/net/lib/sockets/sockets_inet.c` where the other options are set

sayooj-aerlync commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,56 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+				}
+			} 
+			else {
+					NET_ERR("Failed to clone IPv4 multicast packet");
```
Keep the Log priority as NET_DBG, as it may flood the console if errors occur. Let the logs be present only if debug mode is enabled.

sayooj-aerlync replied:
@pdgendt @jukkar Should it go to ` goto err;` if the copy fails?

pdgendt replied:
The packet is already sent, so I don't think we should `goto err;`, but we should increment some error stat count.

pdgendt replied:
I wonder if we need to clone actually, why can't we just call `processing_data(pkt, true)` here?

CC: @jukkar

jukkar replied:
I think we need to clone the packet as the lower levels might have tweaked the data we just sent. This also means that the cloning needs to be done **before** we have actually sent the data.

pdgendt commented on include/zephyr/net/net_context.h:
```diff
@@ -418,7 +418,11 @@ __net_socket struct net_context {
-
+	
+#if defined(CONFIG_NET_IPV4)
+    	bool ipv4_mcast_loop;
```
No spaces before a tab

```suggestion
	bool ipv4_mcast_loop;
```

jukkar replied:
Please add doxygen comment for this newly added value.

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -2150,6 +2153,7 @@ static struct net_pkt *context_alloc_pkt(struct net_context *context,
 					 size_t len, k_timeout_t timeout)
 {
 	struct net_pkt *pkt;
+	
```
Remove unrelated change
```suggestion
```

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1973,6 +1973,21 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
 			}
 
 			break;
+		case IP_MULTICAST_LOOP:
+			if (IS_ENABLED(CONFIG_NET_IPV4)) {
+				if (*optlen != sizeof(int)) {
+					errno = EINVAL;
+					return -1;
+				}
+		
+				*(int *)optval = ctx->ipv4_mcast_loop;
+				*optlen = sizeof(int);
+				
+				return 0;
+			}
+				
+			errno = ENOPROTOOPT;
+			return -1;
```
The `ipv4_mcast_loop` member is compiled out if `CONFIG_NET_IPV4` isn't enabled.

```suggestion
#if defined(CONFIG_NET_IPV4)
		case IP_MULTICAST_LOOP:
			if (*optlen != sizeof(int)) {
				errno = EINVAL;
				return -1;
			}
		
			*(int *)optval = ctx->ipv4_mcast_loop;
			*optlen = sizeof(int);
				
			return 0;
#endif
```

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2591,6 +2606,18 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 			}
 
 			break;
+		case IP_MULTICAST_LOOP:
+			if (IS_ENABLED(CONFIG_NET_IPV4)) {
+				if (optlen != sizeof(int)) {
+					errno = EINVAL;
+					return -1;
+				}
+
+				ctx->ipv4_mcast_loop = *(const int *)optval;
+				return 0;
+			}
+			errno = ENOPROTOOPT;
+			return -1;
```
Same here

```suggestion
#if defined(CONFIG_NET_IPV4)
		case IP_MULTICAST_LOOP:
			if (optlen != sizeof(int)) {
				errno = EINVAL;
				return -1;
			}

			ctx->ipv4_mcast_loop = *(const int *)optval;
			return 0;
#endif
```

Abd002 commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,32 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
 
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET &&
```
ok

Abd002 replied:
thanks for that

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,31 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+		struct net_context *ctx = net_pkt_context(pkt);
+
+		if (ctx && ctx->ipv4_mcast_loop) {
+			struct net_pkt *clone = net_pkt_clone(pkt, K_NO_WAIT);
```
As @jukkar mentioned in some other comment, the packet needs to be cloned before sending it.

Also the formatting for this function is still off.

Abd002 replied:
in formatting u mean the condition before that ?

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1973,6 +1973,19 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
+
+			return 0;
+#endif
+
```
No need for an additional newline

```suggestion
```

Abd002 commented on subsys/net/ip/net_core.c:
```diff
@@ -411,6 +411,31 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+			}
+		}
+	}
+#endif
```
thank you very much, still working on it.

sayooj-aerlync replied:
I would suggest you to get a testbed up first, verify it and then push the code.

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -406,6 +406,30 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+		if (ctx && ctx->ipv4_mcast_loop) {
+			struct net_pkt *clone = net_pkt_clone(pkt, K_NO_WAIT);
+
+			if (clone) {
```
```suggestion
			if (clone != NULL) {
```

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -406,6 +406,30 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+	    net_ipv4_is_addr_mcast((const struct in_addr *)&NET_IPV4_HDR(pkt)->dst)) {
+		struct net_context *ctx = net_pkt_context(pkt);
+
+		if (ctx && ctx->ipv4_mcast_loop) {
```
```suggestion
		if (ctx != NULL && ctx->ipv4_mcast_loop) {
```

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1973,6 +1973,18 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
+				return -1;
+			}
+
+			*(int *)optval = ctx->ipv4_mcast_loop;
```
```suggestion
			*(int *)optval = ctx->ipv4_mcast_loop ? 1 : 0;
```

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2591,6 +2603,16 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+				return -1;
+			}
+
+			ctx->ipv4_mcast_loop = *(const int *)optval;
```
```suggestion
			ctx->ipv4_mcast_loop = *(const int *)optval != 0;
```

pdgendt reviewed (commented):
I'm in favor of explicit typing.

jukkar commented on include/zephyr/net/net_context.h:
```diff
@@ -419,6 +419,11 @@ __net_socket struct net_context {
 
+#if defined(CONFIG_NET_IPV4)
+	/* Controls whether IPv4 multicast packets are looped back to the sender. */
+	bool ipv4_mcast_loop;
```
As I mentioned in IPv6 PR for this feature, we can share this variable between IPv6 and IPv4. So whichever PR is merged 2nd, it needs to change to use union here.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -565,6 +565,9 @@ int net_context_get(sa_family_t family, enum net_sock_type type, uint16_t proto,
 		contexts[i].options.ipv6_v6only = true;
+#endif
+#if defined(CONFIG_NET_IPV4)
+		contexts[i].ipv4_mcast_loop = true;
```
Please take a look of IPv6 PR #87338 and have a Kconfig option to set the initial value.

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -406,6 +406,30 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+					net_pkt_unref(clone);
+				}
+			} else {
+				NET_ERR("Failed to clone IPv4 multicast packet");
```
Please change this to `NET_DBG`

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -406,6 +406,30 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+
+				if (net_recv_data(net_pkt_iface(clone), clone) < 0) {
+					NET_ERR("Failed to process cloned IPv4 packet");
+					net_pkt_unref(clone);
```
Please update the statistics
```
					if (IS_ENABLED(CONFIG_NET_STATISTICS)) {
						net_stats_update_ipv4_drop(net_pkt_iface(pkt));
					}
```

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -406,6 +406,30 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
+				net_pkt_set_iface(clone, net_pkt_iface(pkt));
+
+				if (net_recv_data(net_pkt_iface(clone), clone) < 0) {
+					NET_ERR("Failed to process cloned IPv4 packet");
```
This error print is probably not needed, please remove it.

Abd002 committed:
net: Add support for IPv4_MULTICAST_LOOP in net_context

Add support for the IP_MULTICAST_LOOP socket option to control
multicast packet loopback.

Signed-off-by: AbdElRahman Khalifa <abdelrahman.5alifa@gmail.com>

pdgendt commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1973,6 +1973,18 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
 
 			break;
+#if defined(CONFIG_NET_IPV4)
+		case IP_MULTICAST_LOOP:
+			if (*optlen != sizeof(int)) {
+				errno = EINVAL;
+				return -1;
+			}
+
+			*(int *)optval = ctx->options.ipv4_mcast_loop ? 1 : 0;
+			*optlen = sizeof(int);
+
+			return 0;
```
Similar to #87338 I'd prefer to add a `NET_OPT_IPV4_MCAST_LOOP` to `enum net_context_option` and work with getter/setters here.

pdgendt commented on subsys/net/ip/net_context.c:
```diff
@@ -565,6 +565,10 @@ int net_context_get(sa_family_t family, enum net_sock_type type, uint16_t proto,
 		contexts[i].options.ipv6_v6only = true;
+#endif
+#if defined(CONFIG_NET_IPV4)
+		contexts[i].options.ipv4_mcast_loop = 
```
No trailing spaces

```suggestion
		contexts[i].options.ipv4_mcast_loop =
```

Abd002 replied:
done.

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -1973,6 +1973,19 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
+			ret = net_context_get_option(ctx,
+						     NET_OPT_IPV4_MCAST_LOOP,
+						     optval, optlen);
+
```
Extra empty line could be removed

Abd002 replied:
done.

jukkar commented on subsys/net/lib/sockets/sockets_inet.c:
```diff
@@ -2591,6 +2604,19 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+			ret = net_context_set_option(ctx,
+						     NET_OPT_IPV4_MCAST_LOOP,
+						     optval, optlen);
+
```
same here

Abd002 replied:
done.

pdgendt commented:
It looks like https://github.com/zephyrproject-rtos/zephyr/pull/87338 got in first, will need to rebase this and merge ipv4 with ipv6.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -3685,9 +3716,12 @@ int net_context_set_option(struct net_context *context,
 	case NET_OPT_LOCAL_PORT_RANGE:
 		ret = set_context_local_port_range(context, value, len);
 		break;
+	case NET_OPT_IPV4_MCAST_LOOP:
+		ret = set_context_ipv4_mcast_loop(context, value, len);
+    break;
 	case NET_OPT_IPV6_MCAST_LOOP:
-		ret = set_context_ipv6_mcast_loop(context, value, len);
-		break;
+    ret = set_context_ipv6_mcast_loop(context, value, len);
+    break;
```
Something went wrong with the indentation. You probably do not need to touch the IPv6 variant here.

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -3685,9 +3716,12 @@ int net_context_set_option(struct net_context *context,
 		break;
+	case NET_OPT_IPV4_MCAST_LOOP:
+		ret = set_context_ipv4_mcast_loop(context, value, len);
+    break;
```
Indentation wrong

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -3773,6 +3807,9 @@ int net_context_get_option(struct net_context *context,
 		break;
+	case NET_OPT_IPV4_MCAST_LOOP:
+		ret = get_context_ipv4_mcast_loop(context, value, len);
+    break;
```
ditto

jukkar commented:
Also please rebase the PR and squash the merge commit

Abd002 commented:
shouldn't i merge that to solve this conflict or what  ?

sayooj-aerlync commented:
> shouldn't i merge that to solve this conflict or what  ?

Pull the current zephyr main branch to your fork.
It will show merge conflicts. We will have to solve them.

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -410,6 +410,28 @@ int net_try_send_data(struct net_pkt *pkt, k_timeout_t timeout)
 		goto err;
 	}
 
+#if defined(CONFIG_NET_IPV4)
```
This block is now almost identical to the ipv6, wondering if there'd be a smart way to merge these?

Abd002 replied:
I can make macro and pass function names to it, or I should make something simpler ?

pdgendt replied:
Perhaps a helper function you pass the `pkt` that returns a boolean if the packet should be looped, that way we can split the check logic from the cloning.

Abd002 replied:
sure, I will do it.

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -366,6 +366,35 @@ static inline int check_ip(struct net_pkt *pkt)
+#if defined(CONFIG_NET_IPV4) || defined(CONFIG_NET_IPV6)
+static inline bool process_multicast(struct net_pkt *pkt)
+{
+	struct net_context *ctx = net_pkt_context(pkt);
+	sa_family_t family = net_pkt_family(pkt);
+	if (!ctx) {
+		return false;
+	}
```
```suggestion
	struct net_context *ctx = net_pkt_context(pkt);
	sa_family_t family = net_pkt_family(pkt);
	
	if (ctx == NULL) {
		return false;
	}
```

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -366,6 +366,35 @@ static inline int check_ip(struct net_pkt *pkt)
+
+#if defined(CONFIG_NET_IPV4)
+	if (family == AF_INET) {
+		const struct in_addr *dst = (const struct in_addr *)&NET_IPV4_HDR(pkt)->dst;
+		if (net_ipv4_is_addr_mcast(dst) && net_context_get_ipv4_mcast_loop(ctx)) {
+			return true;
+		}
```
```suggestion
		const struct in_addr *dst = (const struct in_addr *)&NET_IPV4_HDR(pkt)->dst;

		return net_ipv4_is_addr_mcast(dst) && net_context_get_ipv4_mcast_loop(ctx);
```

pdgendt commented on subsys/net/ip/net_core.c:
```diff
@@ -366,6 +366,35 @@ static inline int check_ip(struct net_pkt *pkt)
+#endif
+#if defined(CONFIG_NET_IPV6)
+	if (family == AF_INET6) {
+		const struct in6_addr *dst = (const struct in6_addr *)&NET_IPV6_HDR(pkt)->dst;
+		if (net_ipv6_is_addr_mcast(dst) && net_context_get_ipv6_mcast_loop(ctx)) {
+			return true;
+		}
```
```suggestion
		const struct in6_addr *dst = (const struct in6_addr *)&NET_IPV6_HDR(pkt)->dst;

		return net_ipv6_is_addr_mcast(dst) && net_context_get_ipv6_mcast_loop(ctx);
```

pdgendt reviewed (commented):
Thanks, few more nits

jukkar reviewed (approved):
Thanks! Looks very nice now.
